### DISCUSSION TITLE
Extract and make targets before building a document so targets can be referred to before the end of the pipeline definition

### DISCUSSION THREAD

**Discussion by RichardJActon** [ropensci/targets#652] [Help]
I've been trying out target markdown in a bookdown project and I'm finding
The pattern of writing the entire targets pipeline first, running make then
being able to access the resulting targets a bit difficult as I often find
that I want to be able to refer to the contents of a target somewhere in the 
document before the `tar_make()` call.

A minimal example:

	# Section 1
	
	Here I define the first part of my pipeline
	
	```{targets, thing_number, interactive = TRUE}
	tar_target(Number_of_things, 10)
	```
	
	Now I'd like to show some of information from the part of the pipeline I 
	defined above before continuing the pipeline in section 2
	
	This is a body of text in the document where I'd like to refer to the number of
	things. There are: `r tar_read(Number_of_things)` things. However this does not
	work until after `tar_make()` has been called and the target is in the cache.
	
	# Section 2
	
	```{targets, something_else, interactive = TRUE}
	tar_target(something, "a thing")
	```
	
	# Make {-}
	
	```{r}
	tar_make()
	```

Would it be possible to implement come kind of two step build process where
the target pipeline is extracted from the targets engine chunks in the document
on the first pass, not running any of the R chunks so that tar_make can be run
then the document built after all the targets are generated so that they can be
referenced from R chunks mixed in with the targets engine chunks that define
the pipeline? Or is this already possible and I've missed something, I'm still 
learning targets/target markdown

**Comment by wlandau** [comment]
This was actually what I had first envisioned for Target Markdown. But it would require statically analyzing the all the reports in a pipeline together, then implementing some kind of pre-render step to run the pipeline, and then programmatically modifying the report to print the targets instead of define them, and then render the documents. I decided this would be error-prone in the implementation, confusing for users with edge cases, and hard to generalize to multiple projects. The current system has better alignment with the design and interface of `targets`, and its modular enough (via the language engine) to spread over multiple reports for a pipeline. As a workaround, in most cases, you could check `tar_exist_objects()` before `tar_read()` as you are writing the prose. There are probably other workarounds too.

**Reply by RichardJActon** [reply]
When you say print the targets instead of define them I was envisioning that the targets chunks are simply not run, like setting `eval=FALSE` for the chunk on the Document building pass. My thinking was statically analyse the .Rmd source files only pay attention to the targets engine chunks, building the functions and _targets.R file then run tar_make() followed by the document build which Ignores the targets engine chunks and just runs the R chunks. Nothing fancy like the way interactive mode works with loading the targets into environment when the targets chunks run just ignore running them on the document build pass. But maybe I'm missing how this would not work for some edge cases.

Thanks for the work around suggestion, using it might get a little verbose but I'll give it a go.

**Reply by RichardJActon** [reply]
I'm finding that a pattern where I use `tar_exist_objects()` in the eval arguments of chunks where I want to use those targets works reasonably well e.g.

    ```{r, eval = tar_exist_objects('tar1')}
    tar_read(tar1) # ...
    ```

    ```{r, eval = all(tar_exist_objects(c('tar2','tar3')))}
    tar_read(tar1) # ...
    tar_read(tar2) # ...
    ```

**Reply by RichardJActon** [reply]
So I ran into an issue with dynamic branching and this workaround.

`tar_exist_objects()` returns false if the name you give it is the 'root' of a dynamic branch you need to enumerate for it the specific targets with the randomly generated suffixes. 
`tar_branch_names()` seemed like it might be an option but this only works if you know how many branches you have so you can supply the appropriate indices, is there a function to retrieve all the names of the branches from that branch's 'root' name?
Would it be possible to get `tar_exist_objects()` to play well with `starts_with()` as at the moment it only works with `tar_load()`
Then one could do: `eval = all(tar_exist_objects(starts_with("dynamic_branch_root")))`

**Reply by wlandau** [reply]
You can check if the dynamic target is in `tar_meta()$name`.

**Reply by RichardJActon** [reply]
Ah thanks I didn't think to use `tar_meta` for this. I wrote a convenience function for extracting the children of a given target pattern from `tar_meta`, is this something you'd  like to add, if so I can open a PR?

```
tar_children <- function(target) {
	target <- rlang::ensym(target)
	meta <- tar_meta()
	target_string <- rlang::as_string(target)
	patterns <- meta %>% dplyr::filter(type == "pattern")
	if((!target_string %in% patterns$name) & (target_string %in% meta$name)) {
		stop(paste0(
			target_string, " is a target but not a pattern with children!"
		))
	}
	if(!target_string %in% patterns$name) {
		stop(paste0(
			target_string, " was not found!"
		))
	}
	patterns %>% 
		dplyr::filter(name == rlang::as_string(target)) %>% 
		dplyr::pull(children) %>%
		`[[`(1)
}
```

**Reply by wlandau** [reply]
It's a nice thought, but it feels a bit situation-specific to bake directly into core `targets`.
